lib/nxt_pipeline/pipeline.rb

```diff
@@ -1,81 +1,97 @@
+      name = name.to_sym
+      raise StandardError, "Already registered step :#{name}" if registry[name]
+
+      registry[name] = constructor
+
+      return unless opts.fetch(:default, false)
+      default_constructor ? (raise ArgumentError, 'Default step already defined') : self.default_constructor = constructor
     end
 
-    def failed?
-      failed_step.present?
+    def step(type = nil, **opts, &block)
+      constructor = if block_given?
+                   # make first argument the to_s of step if given
+                   opts.merge!(to_s: type) if type && !opts.key?(:to_s)
+                   block
+                 else
+                   if type
+                     registry.fetch(type) { raise KeyError, "No step :#{type} registered" }
+                   else
+                     default_constructor || (raise StandardError, 'No default step registered')
+                   end
+                 end
+
+      steps << Step.new(constructor, **opts)
     end
 
-    class << self
-      def pipe_attr(name)
-        @pipe_attr_name = name
+    def execute(arg, &block)
+      reset_log
+      configure(&block) if block_given?
+      steps.inject(arg) do |argument, step|
+        execute_step(step, argument)
       end
+    end
 
-      def step(name)
-        self.steps << name
-      end
+    def on_errors(*errors, &callback)
+      error_callbacks << ErrorCallback.new(errors, callback)
+    end
 
-      def rescue_errors(*errors, &block)
-        @rescueable_errors = errors
-        self.rescueable_block = block
-      end
-      
-      def before_each_step(*filters, &block)
-        set_callback :each_step_pipe_through, :before, *filters, &block
-      end
-      
-      def after_each_step(*filters, &block)
-        set_callback :each_step_pipe_through, :after, *filters, &block
-      end
-      
-      def around_each_step(*filters, &block)
-        set_callback :each_step_pipe_through, :around, *filters, &block
-      end
-      
-      attr_reader :pipe_attr_name
-      attr_accessor :rescueable_block
-      
-      def steps
-        @steps ||= []
-      end
-      
-      def rescueable_errors
-        @rescueable_errors ||= []
-      end
+    alias :on_error :on_errors
+
+    def configure(&block)
+      block.call(self)
     end
 
     private
 
-    attr_reader :pipe_attr
+    attr_reader :error_callbacks, :registry
+    attr_accessor :steps, :current_step, :default_constructor
+    attr_writer :log
 
-    def extract_pipe_attr_from_init_params(*attrs)
-      raise ArgumentError, 'You need to pass a keyword param as argument to #new' unless attrs.first.is_a?(Hash)
-      @pipe_attr = attrs.first.fetch(self.class.pipe_attr_name)
-    end
+    def execute_step(step, arg)
+      self.current_step = step.to_s
+      result = step.execute(arg)
 
-    def handle_segment_burst(error, step)
-      @failed_step = step.name.split('::').last.underscore
+      if result # step was successful
+        log[current_step] = 'success'
```

Comment:
Maybe it makes sense to use symbols instead of strings?

Reply:
I mean it's ment to be readable. That'S the only thing I care about at the moment.

Reply:
I agree, I would also prefer to use symbols.

Reply:
@rkallensee @nsommer But it does not make sense to use symbols. It's a human description. When there is an error it contains a sentence.